resolvers/node/resolution.rs

```diff
@@ -61,37 +61,57 @@ use crate::NpmPackageFolderResolver;
 use crate::PackageJsonResolverRc;
 use crate::PathClean;
 
-pub static DEFAULT_CONDITIONS: &[&str] = &["deno", "node", "import"];
-pub static REQUIRE_CONDITIONS: &[&str] = &["require", "node"];
-static TYPES_ONLY_CONDITIONS: &[&str] = &["types"];
-
-#[allow(clippy::disallowed_types)]
-type ConditionsFromResolutionModeFn = crate::sync::MaybeArc<
-  dyn Fn(ResolutionMode) -> &'static [&'static str] + Send + Sync + 'static,
->;
-
-#[derive(Default, Clone)]
-pub struct ConditionsFromResolutionMode(Option<ConditionsFromResolutionModeFn>);
+pub static DEFAULT_CONDITIONS: &[Cow<'static, str>] = &[
+  Cow::Borrowed("deno"),
+  Cow::Borrowed("node"),
+  Cow::Borrowed("import"),
+];
+pub static REQUIRE_CONDITIONS: &[Cow<'static, str>] =
+  &[Cow::Borrowed("require"), Cow::Borrowed("node")];
+static TYPES_ONLY_CONDITIONS: &[Cow<'static, str>] = &[Cow::Borrowed("types")];
+
+#[derive(Debug, Clone)]
+pub struct ConditionResolverOptions {
+  pub conditions: Vec<Cow<'static, str>>,
+  pub default_import_conditions: Vec<Cow<'static, str>>,
+  pub default_require_conditions: Vec<Cow<'static, str>>,
+}
 
-impl Debug for ConditionsFromResolutionMode {
-  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-    f.debug_struct("ConditionsFromResolutionMode").finish()
+impl Default for ConditionResolverOptions {
+  fn default() -> Self {
+    Self {
+      conditions: Vec::new(),
+      default_import_conditions: DEFAULT_CONDITIONS.to_vec(),
+      default_require_conditions: REQUIRE_CONDITIONS.to_vec(),
+    }
   }
 }
 
-impl ConditionsFromResolutionMode {
-  pub fn new(func: ConditionsFromResolutionModeFn) -> Self {
-    Self(Some(func))
+#[derive(Debug, Default, Clone)]
+pub struct ConditionResolver {
+  options: ConditionResolverOptions,
+}
+
+impl ConditionResolver {
+  pub fn new(options: ConditionResolverOptions) -> Self {
+    Self { options }
   }
 
-  fn resolve(
+  pub fn resolve(
     &self,
     resolution_mode: ResolutionMode,
-  ) -> &'static [&'static str] {
-    match &self.0 {
-      Some(func) => func(ResolutionMode::Import),
-      None => resolution_mode.default_conditions(),
-    }
+  ) -> Vec<Cow<'static, str>> {
+    let default_conditions = match resolution_mode {
+      ResolutionMode::Import => &self.options.default_import_conditions,
+      ResolutionMode::Require => &self.options.default_require_conditions,
+    };
+    self
+      .options
+      .conditions
+      .iter()
+      .chain(default_conditions)
+      .cloned()
+      .collect()
```

Comment:
Can we change this to do the allocations in the constructor by precomputing the two possible values? (and preferably avoid any allocations there when `self.options.conditions` is empty, though that's not as big of a deal) I think this could return a `&[Cow<'static, str>>]`

Sorry, this might seem nitpicky, but it's not really. I did a lot of work removing allocations from node resolution in order to make it faster.